src/main/scala/viper/gobra/frontend/Desugar.scala

```diff
@@ -651,33 +664,37 @@ object Desugar {
             if (left.size == right.size) {
               sequence((left zip right).map{ case (l, r) =>
                 for{
-                  le <- unit(in.Assignee.Var(assignableVarD(ctx)(l)))
                   re <- goE(r)
+                  le <- unit(in.Assignee.Var(getVar(l)(re.typ)))
                 } yield singleAss(le, re)(src)
               }).map(in.Seqn(_)(src))
             } else if (right.size == 1) {
               for{
-                les <- unit(left.map{l => in.Assignee.Var(assignableVarD(ctx)(l))})
                 re  <- goE(right.head)
+                les <-
+                  unit(left.map{l =>  in.Assignee.Var(getVar(l)(typeD(info.typ(l), Addressability.exclusiveVariable)(src)))})
               } yield multiassD(les, re)(src)
             } else { violation("invalid assignment") }
 
           case PVarDecl(typOpt, right, left, addressable) =>
 
-            if (left.size == right.size) {
-              sequence((left zip right).map{ case (l, r) =>
+            if (left.size == right.size && left.size == addressable.size) {
+              sequence((left zip right zip addressable).map{ case ((l, r), addr) =>
                 for{
-                  le <- unit(in.Assignee.Var(assignableVarD(ctx)(l)))
                   re <- goE(r)
+                  addressability = if(addr) Addressability.Shared else Addressability.Exclusive
```

Comment:
I can probably reverse these changes and just used Addressability.Exclusive because Addressability does not seem to be used anywhere, it is in fact ignored when getVar creates a new variable